Add BME680 via BSEC integration

trvrnrth opened the pull request:
## Description:
Adds new sensor components integrating the BME680 Temperature+Pressure+Humidity+Gas Sensor via BSEC providing additional Indoor Air Quality (IAQ), CO2 equivalent and Breath Volatile Organic Compounds (VOC) equivalent measurements.

**Related issue (if applicable):** fixes https://github.com/esphome/feature-requests/issues/329

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#793

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

trvrnrth committed:
Add BME680 via BSEC integration

OttoWinter commented on esphome/components/bme680_bsec/__init__.py:
```diff
@@ -0,0 +1,44 @@
+
+CONFIG_SCHEMA = cv.Schema({
+    cv.GenerateID(): cv.declare_id(BME680BSECComponent),
+    cv.Optional(CONF_TEMPERATURE_OFFSET, default=0): cv.float_range(min=-100, max=100),
```
Is this in celsius? If so, please use `cv.All(cv.temperature, cv.float_range(min=-100, max=100))`

trvrnrth replied:
It is. Will do.

trvrnrth replied:
Amended in 3abfc3c to just use cv.temperature as there are no specific restrictions on range that I'm aware of (other than common sense!).

OttoWinter commented on esphome/components/bme680_bsec/bme680_bsec.cpp:
```diff
@@ -0,0 +1,218 @@
+      // First state update when IAQ algorithm achieves full calibration
+      (this->state_save_counter_ == 0 && this->get_iaq_accuracy_() >= 3)
+      // Then update every state save interval minutes
+      || ((this->state_save_counter_ * this->state_save_interval_ * 60 * 1000) < millis())) {
```
Is this code right? My checking `< millis()` this will execute only once every 40 days (when millis rolls over)

trvrnrth replied:
It performs as I'd expect in my testing at least. The interval (in minutes) is multiplied out to milliseconds, and also by the counter. If we're doing a save we increment the counter immediately so the next check will wait for millis() to catch up again.

I *think* both sides will overflow in the same manner and thus all will be fine, but I should perhaps double check that assumption. FWIW this was lifted from the Bosch example at https://github.com/BoschSensortec/BSEC-Arduino-library/blob/996a03594faf3b889efd69a0ca92e55a9ce242b2/examples/basic_config_state/basic_config_state.ino#L164

trvrnrth replied:
Further testing suggests this is fine, but after further thought I've simplified the approach and mirrored that taken by espurna in https://github.com/xoseperez/espurna/blob/aebefdf941b30d3db57c1dda4f2a2f7ea8e9d8c4/code/espurna/sensors/BME680Sensor.h#L240

There's no longer an initial save outside of the normal interval when full calibration is first reached which keeps things simpler. All interval based saves will wait for full calibration to be reached which seemed like a reasonable idea as that's what you'll be wanting to start from on restart.

There are various different approaches that could be taken here, but on balance simplicity and parity with espurna seemed like a decent choice.

OttoWinter commented on platformio.ini:
```diff
@@ -20,6 +20,7 @@ lib_deps =
     1655@1.0.2  ; TinyGPSPlus (has name conflict)
     6865@1.0.0  ; TM1651 Battery Display
     6306@1.0.3  ; HM3301
+    bsec=https://github.com/trvrnrth/BSEC-Arduino-library
```
Hmm, unfortunately this is problematic with the ESPHome license (GPL for cpp code).

Adding this line would mean that anybody who shares a debug esphome build would violate the license.

Please remove this line from platformio.ini and wrap all the cpp code files in `#ifdef USING_BSEC` and add a `cg.add_build_flag("-DUSING_BSEC")` in the python code.

trvrnrth replied:
The other problem is that I was rather hoping to get https://github.com/BoschSensortec/BSEC-Arduino-library/pull/51 merged first so it doesn't depend on my fork. But that's a a separate problem.

I knew the licensing issue would crop up but wasn't sure what best to do about it other than call it out in the docs. The reason I added this line was so that the linters would run as otherwise they could not find `bsec.h`. Would removing it not break them? My understanding was that for an actual build this config wasn't used and instead the `cg.add_library` call would be used, gated by explicit configuration of the platform.

OttoWinter replied:
Correct, the linter will not accept it with the BSEC library missing. But with the `#ifdef USING_BSEC`, the linter will skip linting because that symbol is only defined when loaded through a user-install (not from the "fake" pio project in platformio.ini)

trvrnrth replied:
I guess I'd been thinking that as the only place the "fake" pio project is used is for the purposes of linting and in that scenario no build including libalgobsec.a is produced or distributed that there wouldn't be a licensing issue. Or have I misunderstood?

OttoWinter replied:
True, it's not distributed anywhere (right now at least). And probably it's ok as-is. I don't know the license (and all its precedent cases) well enough to confidently say there are no issues with it though, so  I would prefer if we keep on the safe side and only include this for all developers.

(Also not sure at what point the BSEC license applies, after downloading or when linked to a project. If the former that would also be a reason).

trvrnrth replied:
That seems fair. I've done this in dab3e64. Are you happy to keep the tests in `test1.yaml`?

OttoWinter reviewed (commented):
Thanks for adding this, I added some comments about the code (and a note about the licensing issues)

trvrnrth committed:
Use temperature validator for offset

trvrnrth committed:
Gate BSEC-dependant code behind define

This is a safer approach to licensing and reduces download of BSEC library only to scenarios in which the component is specifically configured.

trvrnrth committed:
Simplify state save logic

This borrows the approach taken by espurna, with saves performed as soon after each interval that full calibration is achieved. Always having a fully calibrated state to restart from seems more sensible.

There is no longer an initial save when full calibration is first reached, but I favour the simplicity over this minor nicety.

trvrnrth committed:
Add setup log line for parity with other sensors

SenexCrenshaw commented:
@trvrnrth I just started a project with a bme680 and wanted to see what this does. I tried this PR and the check_status never comes back ok. the bme_sec piece becomes OK but the bmestatus part of the check never does. Any thoughts? My sensor works fine with the regular driver.

trvrnrth committed:
Use latest official BSEC Software Library release

This now includes builds of the library that compile for esp8266 without need for modification of the core platform linker rules.

trvrnrth committed:
Merge branch 'dev' into bme680-bsec

trvrnrth commented:
Opened for review now that the latest official release of the BSEC library contains builds of the library that compile for esp8266 without need for modification of the core platform linker rules. This removes the need for the dependency on my temporary fork.

kbx81 commented:
This is cool -- I've got a few of these sensors on hand. Can't wait to try it out. 😄

trvrnrth committed:
Only publish new text sensor states if changed

This cuts down on unnecessary chatter from the IAQ accuracy text sensor.

trvrnrth committed:
Merge branch 'dev' into bme680-bsec

trvrnrth committed:
Update to v1.0.0

https://github.com/trvrnrth/esphome-bsec-bme680/releases/tag/v1.0.0

Includes:
- Optional numeric IAQ sensor
- Configurable sample rate (LP or ULP)
- Tidied up yeilding to resolve issues with TCP stack servicing/API
  disconnects

trvrnrth committed:
Reduce RAM requirements for setup

These changes only intitialize the config arrays when necessary rather
than keeping all options hanging around at all times.

trvrnrth committed:
Update last save storage to adhere to lint rules

trvrnrth committed:
Change include indenting to match lint rules

trvrnrth committed:
Remove usage of Bsec helper class

This removes usage of the Bosch provided Bsec helper class and
instead interfaces more directly with the underlying library.
Most importantly this allows removal of the forced delay within
the component loop which was causing various problems within the
ESPHome ecosystem.

Unfortunately due to the added timing complexities this does mean
that I have removed support for reading from multiple sensors as it
would take considerable extra effort to appropriately manage the
bsec library state.

trvrnrth committed:
Lint and move tests to test4 as test1 is full

rdehuyss commented:
Thanks to everyone for all the hard work on this PR. What still needs to be done to have it merged?

kbx81 commented:
I was going to test this out this weekend but seem to be running into some issues with the config validation:

```
ERROR Unable to load component bme680_bsec.sensor:
Traceback (most recent call last):
  File "/Users/kburzinski/Development/Embedded/ARM/Espressif/esphome-contrib/trvrnrth/esphome/esphome/config.py", line 161, in _lookup_module
    module = importlib.import_module(f"esphome.components.{domain}")
  File "/Users/kburzinski/.pyenv/versions/3.7.9/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "~/esphome-contrib/trvrnrth/esphome/esphome/components/bme680_bsec/sensor.py", line 34, in <module>
    sensor.sensor_schema(UNIT_CELSIUS, ICON_THERMOMETER, 1),
TypeError: sensor_schema() missing 1 required positional argument: 'device_class_'
INFO Detected timezone 'CST' with UTC offset -6 and daylight savings time from 13 March 02:00:00 to 06 November 02:00:00
Failed config
```

I tried doing a little patching on my own and got the config to validate, but then ran into a mess of other compiler errors. I'm not quite sure what's up -- perhaps it needs to be rebased and updated for current `dev`?

trvrnrth committed:
Merge branch 'dev' into bme680-bsec

trvrnrth committed:
Black and add device classes

trvrnrth commented:
@kbx81 I've formatted with black and added in the missing device classes. Seems fine to me now so not sure what other problems you were running into.

OttoWinter commented on tests/test4.yaml:
```diff
@@ -56,6 +56,12 @@ time:
 tuya:
   time_id: sntp_time
 
+bme680_bsec:
```
Sorry to request another change, but we need to remove this line from the tests.

By using this component (and downloading the bsec code), the user agrees with the Bosch license. That's ok for individual users, but if we put it in the testcases, the whole project agrees to that license - I haven't looked at their license in detail, but it's safer to keep it out of the tests here.

trvrnrth replied:
That's fine. I [rather suspected](https://github.com/esphome/esphome/pull/1313#discussion_r499262374) that would be the stance.

trvrnrth committed:
Remove from tests due to licensing restrictions

kbx81 commented:
> @kbx81 I've formatted with black and added in the missing device classes. Seems fine to me now so not sure what other problems you were running into.

Sweet! Thank you...that helped. Now I seem to be running into some compilation errors...which I believe is what was happening yesterday:

```
src/main.cpp:90:1: error: 'bme680_bsec' does not name a type
 bme680_bsec::BME680BSECComponent *bme680_bsec_bme680bseccomponent;
 ^
 .....
~/esphome-configs/kbx/shared/esp_thermostat_nextion_common_config.yaml: In function 'void setup()':
~/esphome-configs/kbx/shared/esp_thermostat_nextion_common_config.yaml:2355:3: error: 'bme680_bsec_bme680bseccomponent' was not declared in this scope
~/esphome-configs/kbx/shared/esp_thermostat_nextion_common_config.yaml:2355:41: error: 'bme680_bsec' does not name a type
Compiling .pioenvs/esp_thermostat_dev_7inch/libbf7/ESPAsyncWebServer-esphome/WebRequest.cpp.o
Compiling .pioenvs/esp_thermostat_dev_7inch/libbf7/ESPAsyncWebServer-esphome/WebResponses.cpp.o
~/esphome-configs/kbx/shared/esp_thermostat_nextion_common_config.yaml:681:49: error: 'bme680_bsec' has not been declared
~/esphome-configs/kbx/shared/esp_thermostat_nextion_common_config.yaml:682:52: error: 'bme680_bsec' has not been declared
```

Is it possible I've left something out of the config?

trvrnrth commented:
@kbx81 Strange. There's nothing particularly special about the config. The only slight oddity compilation wise is that all the component code is gated behind a define to make things clearer/safer with regard to licensing requirements. But that should get added for you by the code-gen. You should be able to just use the minimal example in the docs. If you have a non-working config I can perhaps see if I can reproduce?

kbx81 commented:
Happy to share! Here's what I've got config-wise:

```yaml
# BME680 BSEC sensor
bme680_bsec:
  address: 0x77

sensor:
 # BME680 BSEC sensor
  - platform: bme680_bsec
    temperature:
      id: esp_thermostat_bme680_temperature
      name: ${node_friendly_name} BME680 Temperature
      filters:
        - offset: -1.2
      on_value:
        then:
          - lambda: |-
              std::string temperature_string = esp32_thermostat::round_float_to_string(id(esp_thermostat_bme680_temperature).state * 1.8 + 32) + "\xB0";
              id(textSensor1a).set_state(temperature_string, false, true);
              esp32_thermostat::update_climate_table_temperature(id(climate_table_local_row), id(esp_thermostat_bme680_temperature).state);
    humidity:
      id: esp_thermostat_bme680_humidity
      name: ${node_friendly_name} BME680 Humidity
      on_value:
        then:
          - lambda: |-
              std::string humidity_string = esp32_thermostat::round_float_to_string(id(esp_thermostat_bme680_humidity).state) + "%";
              id(textSensor1b).set_state(humidity_string, false, true);
              esp32_thermostat::update_climate_table_humidity(id(climate_table_local_row), id(esp_thermostat_bme680_humidity).state);
    pressure:
      id: esp_thermostat_bme680_pressure
      name: ${node_friendly_name} BME680 Pressure
      on_value:
        then:
          - lambda: |-
              std::string pressure_string = esp32_thermostat::round_float_to_string(id(esp_thermostat_bme680_pressure).state) + "hPa";
              id(textSensor1c).set_state(pressure_string, false, true);
    gas_resistance:
      id: esp_thermostat_bme680_gas_resistance
      name: ${node_friendly_name} BME680 Gas Resistance
      on_value:
        then:
          - lambda: |-
              std::string gas_res_string = esp32_thermostat::round_float_to_string(id(esp_thermostat_bme680_gas_resistance).state);
              id(textSensor1d).set_state(gas_res_string, false, true);
    iaq:
      id: esp_thermostat_bme680_iaq
      name: ${node_friendly_name} BME680 IAQ
      on_value:
        then:
          - lambda: |-
              std::string iaq_string = esp32_thermostat::round_float_to_string(id(esp_thermostat_bme680_iaq).state);
              id(textSensor1d).set_state(iaq_string, false, true);
    co2_equivalent:
      id: esp_thermostat_bme680_eco2
      name: ${node_friendly_name} BME680 eCO2
    breath_voc_equivalent:
      id: esp_thermostat_bme680_evoc
      name: ${node_friendly_name} BME680 Breath eVOC

text_sensor:
  - platform: bme680_bsec
    iaq_accuracy:
      id: esp_thermostat_bme680_accuracy
      name: ${node_friendly_name} Accuracy
```

The various `on_value` actions I have update text sensors on a display I'm using with @SenexCrenshaw's `nextion_upload` branch. (I've merged your branch here with his `nextion_upload` and `sgp40` branches so I can test them all together.)

kbx81 commented:
I got it to build! I added `build_flags: -DUSING_BSEC` to my config to force`USING_BSEC` to be defined. Not sure why it's not being automagically added in this case.

If you'd like a closer look, I pushed up the full configuration I've got for this to my config repo [here](https://github.com/kbx81/esphome-configs/tree/nextion-csc). 😄

kbx81 commented:
I just also want to add that this is _wayyyyyy_ cool. Adds a lot of capability for this little sensor. Love it! 😍

trvrnrth commented:
@kbx81 Glad to hear the component itself is working! It's really odd that the build flag wasn't added for you though. If you build with verbose logging on do you see the `DEBUG Adding build flag: -DUSING_BSEC` log line?

OttoWinter reviewed (approved):
Thanks! LGTM :+1:

kbx81 commented:
> @kbx81 Glad to hear the component itself is working! It's really odd that the build flag wasn't added for you though. If you build with verbose logging on do you see the `DEBUG Adding build flag: -DUSING_BSEC` log line?

Ah ha! I believe #1651 will fix the issue I was having. I have manually specified `build_flags` to enable the PSRAM on the WROVER module in my design and I believe it's flattening what you were adding in your code. 😄

Shout out to @SenexCrenshaw for helping with identifying this. 😄🍻
